### ISSUE TITLE
Set default map-matching config to optimal values found in Reporter Quality Testing Rig

### ISSUE THREAD

**Issue by drewda** [opentraffic/reporter#88]
Based on @mxndrwgrdnr's analysis [optimizing map-matching parameters](http://nbviewer.jupyter.org/github/opentraffic/reporter-quality-testing-rig/blob/6e822b1091b636f24da66fa944e3a6e53aef3487/notebooks/validation.ipynb#4.-Optimize-Map-Matching-Parameters) let's set reasonable default values for the primary map-matching parameters that are baked into the default Docker images:

- sigmz_z
- beta
- max_route_time_factor

Any other [default parameters](https://github.com/valhalla/valhalla/blob/master/docs/meili/configuration.md#map-matching-parameters) worth tweaking?

(Ideally as part of deploying Reporter instances, a data provider will make use of the Reporter Quality Testing Rig to come up with the ideal set of parameters to use given their GPS update frequency, typical GPS accuracy, and local road network density -- but if they don't go through the entire optimization exercise, we should at least give them decent defaults to begin with.)

**Comment by kevinkreiser** [comment]
@mxndrwgrdnr can you list your thoughts on the parameters above, it seems like you'd be the best equipped to answer this. i'll happily set the defaults accordingly

**Comment by mxndrwgrdnr** [comment]
Based on my analyses I would recommend the following defaults:
- sigma_z: 7
- beta: 2
- max_route_time_factor: 0.5

**Comment by mxndrwgrdnr** [comment]
Am I interpreting the "max_route_time_factor" correctly as essentially the same measure as my "speed error threshold"?

**Comment by kevinkreiser** [comment]
its a multiplier against the real time between 2 gps measurements. so maybe you mean 1.5? a value of 1.5 would allow for paths that take 1.5 times longer (a la the speeds stored in the graph edges) than the time in the actual gps readings between two gps points. So if you give me gps and it has 2 points a and b, and the timestamps on them have a delta of 5 seconds, then a value of 1.5 will allow routes between a and b of up to 7.5 seconds in length as determined by the default (or tagged) speed information stored in the routing graph. Any paths greater than 7.5 seconds will not be considered as candidates.

i've tried to explain this in multiple ways above to make it less confusing, maybe one of the ways will make sense? maybe it will just be more confusing... but anyway a value of .5 would mean that we would reject any paths between a given pair of gps points that took any longer than half the actual time elapsed between the two points. in other words we basically should never allow values lower than 1 and we should probably really never allow values lower than 1.5. doing so will ensure that we will have more discontinuities in our matches. i suppose if we dont want to keep matches that are faster than the data says you can drive, then this would also be a way to do it.

**Comment by mxndrwgrdnr** [comment]
OK I understand. I think based on Drew's comment above, he was interpreting this parameter as the equivalent to my "speed error threshold", which, under that interpretation, would be designed to toss out matches that are faster than the data says you can drive. Can you confirm @drewda?

**Comment by drewda** [comment]
So `max_route_time_factor=1.5` would eliminate matches that need to go more than 150% of OSM speed limits to traverse candidate segments between two points?

If so, that makes sense to me -- and seems like a reasonable default cutoff for (likely erroneous) high speeds.

**Comment by mxndrwgrdnr** [comment]
No, I think what Kevin is saying is that a `max_route_time_factor` of 1.5 would eliminate matches that take longer than 1.5x the measured elapsed time to traverse. In other words, to eliminate matches based on _excessive_ speeds, which was the primary concern of WB, we would need a `max_route_time_factor` < 1.

**Comment by drewda** [comment]
Ah, O.K., sounds like I flipped the base!

In any case, I think we're all in sync that the goal is to have Reporter skip candidate matches that involve speeds very well in excess of the OSM speed limits.

**Comment by mxndrwgrdnr** [comment]
So in that case, we're really looking for a `min_route_time_factor`, which to my knowledge does not yet exist. @kevinkreiser?

**Comment by kevinkreiser** [comment]
@mxndrwgrdnr @drewda well darn it i guess all my explaining only muddied the waters. sorry about that. drews first assertion was correct.

maybe explaining it in terms of the algorithm will help. we take the time between the two gps points and multiply it by `max_route_time_factor` to get a `max_allowed_path_time`. this is the maximum amount of time any path between two points can take as determined by OSM speeds. then when we are expanding edges in the graph to find paths from the origin candidate to `n` destination candidates, we wont consider any edges that would have a path time greater than `max_allowed_path_time`. what this means is that we prune edges from the search (stops the search at that edge) when a given edge would have taken too long to get to based on OSM road speeds.

this means that a `max_route_time_factor` of `1.5` would allow the OSM road speeds to be 66% that of the speed you were traveling. remember that once we have the path we throw out the speed information from OSM and apply the timestamps from your gps back to it. So concretely:

your gps had 2 points in it. everything we do is between 2 points. the 2 points say you took 10 seconds to travel 100 meters (bee line between them). say `max_route_time_factor` is `1.5`. we'll allow any paths between the candidates that are 15 seconds or less. now say we find a path and the actual distance along the geometry of the route network is 150 meters. going by this we now know that between those two points you were traveling at 150/10=15 meters per second. Notice how in the end we didnt care what OSM speeds thought you were doing over that 150 section of road. But because of the `max_route_factor` we know that the OSM speeds must have been at least 15 / 1.5 = 10 meters per second (weighted average over all lengths of all edges in the path).

i hope this makes sense :smile:

**Comment by mxndrwgrdnr** [comment]
OK, now all makes sense. Let's go with the following default parameters for now:
- sigma_z: 7
- beta: 2
- max_route_time_factor: 1.5

**Comment by kevinkreiser** [comment]
so i ran some tests with these values and i noticed that a `sigma_z` of 7 seems to have some problems for certain routes. specifically this happens:

![bildschirmfoto vom 2017-08-09 08-47-43](https://user-images.githubusercontent.com/697548/29122209-6f96186a-7cdf-11e7-9902-b5031c293c36.png)

the blue line is the input the green line is the correct path and the red line is the match using the above parameters. in this case the only parameter that affected this route was `sigma_z` the others had no impact

**Comment by kevinkreiser** [comment]
i am wondering if we want to be a bit more conservative with `sigma_z` and `max_route_time_factor` maybe `5` and `2` respectively? i'm making changes to the reporter to make these values easily changable for a given container build (they'll be at the top of the dockerfile) so you can play around with the values

**Comment by kevinkreiser** [comment]
similarly a `max_route_time_factor` of even `2` shows this where the paths take about the same time and the noise ceiling is large enough to encompass paths which are far away but have the same time to traverse:

![bildschirmfoto vom 2017-08-09 09-07-05](https://user-images.githubusercontent.com/697548/29123104-96dabb62-7ce2-11e7-9a88-55740e4ef656.png)

im not sure that this argues against setting the `max_route_time_factor` lower but it instead may argue not to match on routes with such coarse accuracy (unless the underlying road network is very sparse)

**Comment by mxndrwgrdnr** [comment]
If those values are producing visibly better results then by all means go with them. It is perplexing to me, though, since I've yet to produce a case in my own testing that looks like these examples, where the "matched" route is not at all reasonable. Most of what I've seen in terms of mismatches have all been in the realm of "that's a mismatch but I can't expect the algorithm to do much better given how bad the data looks". Both of these examples are cases where we would clearly expect the algorithm to do better.

I haven't, for example, seen a case where map matching has actually skipped over the correct route in favor of an erroneous one (i.e. the true route is physically located "between" the measurements and the erroneous match) such as occurs in the top of your first image above. The only reason I can think for that happening is that the algorithm is bumping up against some hard-coded thresholds (`max_route_time_factor`, etc.) that causes it to toss out the correct route. In that image, the correct route, which appears to be on a secondary road, is getting tossed out in favor of a route that lies exclusively on a major throughway, even though it is clearly a much worse match geometrically. My guess is that this is attributable to differences in the estimated speeds.

Similarly, the second image seems to go haywire in a section of the route that goes through a cemetery. Even if people _are_ driving here, the speed estimates for those road segments are probably way off (just a guess). The matched route is clearly so much farther away from the measurements than the correct route (as well as multiple other routes for that matter) that I have to believe something else is going on in these cases besides the algorithm parameters, although if changing the params helps then its hard to argue that case. How frequently are you seeing this pattern?

Anyways, if you are clearly getting better results from different parameters then go with those. We (I) will want to dig into why that's the case, though.

**Comment by mxndrwgrdnr** [comment]
So, maybe I'm barking up the wrong tree here, but I dug a little deeper into the first instance and its a pretty weird one. Technically the routes are on the same streets (Cartesiusweg) but they have different geometries in OSM, [one of which](https://www.openstreetmap.org/way/7060563#map=18/52.10163/5.08827) is classified as a highway and has a speed limit and [one of which](https://www.openstreetmap.org/way/22914514#map=18/52.10163/5.08693) which does not have a road classification or speed limit. Could this have something to do with the mismatch? 

Similarly, the second case it looks like the route goes through a [segment](https://www.openstreetmap.org/way/53608617) that is classified as a "track" and has no speed tags. The mismatch in this case is also a "track" so this doesn't totally explain why it was chosen over the true route, but still I wonder how good we can expect the algorithm to perform in this instance?

Do you have any other cases of this phenomenon occurring on road segments that stick to "normal" road types?

**Comment by kevinkreiser** [comment]
@mxndrwgrdnr  first i should have mentioned a caveat, the data used in generating these is over a year old, so there are definitely discrepancies in tags from then vs now. 

let me explain a little more, how this process works. i have a bunch of random tests (80 of them). the parameters above were set to 3, 4.07 and 5 for beta, sigma z and time factor respectively. in the first example, when i changed all of them to the suggested values, it failed, so i changed them all back but kept the suggested sigma z and got the result above. i was trying to single out what parameter made the change.

similarly for the second one, with all of them changed it failed, i narrowed it down to the time factor being the one that exposes this failure.

i'm not trying to cast doubt on methods used to get these parameters or anything just saying that they get worse results for these routes. it may very well be the case that indeed we lost these 2 routes but actually we do better on hundreds, and in that case, its totally worth it. i just dont know how to measure that though.

basically the training set that lead to these values is slightly different than my training set and i can see we get different results. so what im wondering is how do we come up with a training set that accurately represents real data we will get. im increasingly starting to think, that the answer is we need to run real data through it and have a human look at the results (as i did in the early days). but i think the numbers you've arrived at are at the very least a good start even if based on simulated data.

boy im not sure if ive answered any of your questions at this point!

**Comment by mxndrwgrdnr** [comment]
Agreed, if we are doing better on hundreds then 100% go with your values. In these two instances the matches were just _so_ wrong that it seemed like something else must be going on here. I also agree that if these are the types of matches we're getting then we aren't going to be able to do much better than the eyeball test.

Manually inspecting the results of real data we'd definitely be able to catch mismatches such as the ones you've posted here. We wouldn't, however, really be able to catch mismatches that fall into the that's-a-mismatch-but-I-can't-expect-the-algorithm-to-do-any-better" category, where the match looks reasonable given the data. Does that make sense? I've been focused on the latter since I thought we had the former kind of under control, but that no longer seems to be the case. So right now I'm mostly concerned with why I haven't been seeing these kinds of mismatches and now you are.

**Comment by kevinkreiser** [comment]
yeah im not sure exactly what is going on, so i just thought further investigation might be warrented. just to be clear i wasnt suggesting that my values were doing better on hundreds i was suggesting that maybe your suggested values do and that there just happen to be 2 in my set of routes that dont work. 

i totally agree, catching that second case by hand isn't really possible to fix. as to why you havent seen them, id say it has to do with our testing methodologies being different. we can go over these when you are back and see if we can reason out what can be done. i think the most likely bit is that my testing sets a noise ceiling and uses that ceiling as the accuracy measure passed to map match. so if i set it at 100m and then the noise ends up being more like 20m (randomly it can be) then the algorithm will still include 100m worth of candidates around a given point. the problem is, im not sure what real gps will do. i can say ive seen wildly large accuracy values in the real data (50km). i thought the algorithm sought to minimize the difference in the trace and the path taken and that that would account for this though so im not sure. i just wanted to raise it to be sure we all knew about it

**Comment by dnesbitt61** [comment]
These were set in the Reporter Docker settings.
